File: crow-cli/src/session.rs
//! Agent session management — persistence layer for conversation state.
//!
//! Port of Python crow-cli/agent/session.py.
//! agent_id = "{session_id}-{agent_idx}" is the PK.

use std::path::Path;

use crate::config::Config;
use crate::coolname;
use crow_memory_sdk::MemoryClient;

/// Manages conversation state and persistence via LanceDB.
pub struct AgentSession {
    pub agent_id: String,
    pub session_id: String,
    pub agent_idx: i64,
    pub cwd: String,
    pub messages: Vec<serde_json::Value>,
    pub model_identifier: String,
    pub tools: serde_json::Value,
    #[allow(dead_code)]
    pub request_params: serde_json::Value,
    #[allow(dead_code)]
    pub prompt_id: String,
    #[allow(dead_code)]
    pub prompt_args: serde_json::Value,
}

impl AgentSession {
    /// Add a message to in-memory list AND persist to database.
    pub async fn add_message(
        &mut self,
        store: &MemoryClient,
        msg: serde_json::Value,
        usage: Option<serde_json::Value>,
    ) {
        self.messages.push(msg.clone());
        if let Err(e) = store
            .add_message(&self.agent_id, &msg, usage.as_ref())
            .await
        {
            tracing::warn!("failed to persist message: {e}");
        }
    }

    /// Build and persist an assistant response.
    pub async fn add_assistant_response(
        &mut self,
        store: &MemoryClient,
        thinking: &[String],
        content: &[String],
        tool_call_inputs: &[serde_json::Value],
        usage: Option<serde_json::Value>,
    ) {
        if content.is_empty() && tool_call_inputs.is_empty() {
            return;
        }
        let thinking_text = thinking.join("");
        let content_text = content.join("");
        let mut msg = serde_json::json!({
            "role": "assistant",
            "content": content_text,
        });
        if !thinking_text.is_empty() {
            msg["reasoning_content"] = serde_json::Value::String(thinking_text);
        }
        if !tool_call_inputs.is_empty() {
            msg["tool_calls"] = serde_json::Value::Array(tool_call_inputs.to_vec());
        }
        self.add_message(store, msg, usage).await;
    }

    /// Add tool results to the conversation.
    pub async fn add_tool_response(
        &mut self,
        store: &MemoryClient,
        tool_results: &[serde_json::Value],
    ) {
        for result in tool_results {
            self.add_message(store, result.clone(), None).await;
        }
    }

    /// Load an existing session from the database.
    #[allow(dead_code)]
    pub async fn load(store: &MemoryClient, agent_id: &str) -> anyhow::Result<Self> {
        let agent = store
            .get_agent(agent_id)
            .await?
            .ok_or_else(|| anyhow::anyhow!("agent '{agent_id}' not found"))?;
        let messages = store.load_messages(agent_id).await?;
        Ok(Self {
            agent_id: agent.agent_id,
            session_id: agent.session_id,
            agent_idx: agent.agent_idx,
            cwd: agent.cwd,
            messages,
            model_identifier: agent.model_identifier,
            tools: agent.tool_definitions,
            request_params: agent.request_params,
            prompt_id: agent.prompt_id,
            prompt_args: agent.prompt_args,
        })
    }
}

/// The session's CURRENT conversation generation for resume: the highest
/// agent_idx, most recent created_at on ties.
///
/// Compaction starts a new generation (idx+1) whose history is self-contained
/// (system + summary + everything since). Resuming from ANY older generation
/// resurrects superseded history — the model sees stale duplicated context
/// and starts repeating itself — so resume must always pick the chain head.
pub fn pick_resume_agent(agents: &[crow_memory_sdk::AgentRecord]) -> &crow_memory_sdk::AgentRecord {
    agents
        .iter()
        .max_by(|a, b| {
            a.agent_idx
                .cmp(&b.agent_idx)
                .then_with(|| a.created_at.cmp(&b.created_at))
        })
        .expect("list_agents returned empty")
}

/// Create a new agent session with system prompt, persist to DB.
pub async fn make_agent_session(
    config: &Config,
    store: &MemoryClient,
    tools: &serde_json::Value,
    model_id: &str,
    cwd: &str,
    session_id: Option<&str>,
    agent_idx: Option<i64>,
) -> anyhow::Result<AgentSession> {
    // Load template
    let template = if !config.system_prompt.is_empty() {
        config.system_prompt.clone()
    } else {
        let template_path = config.config_dir.join("prompts/system_prompt.hbs");
        std::fs::read_to_string(&template_path).unwrap_or_default()
    };

    let prompt_id = store
        .lookup_or_create_prompt(&template, "crow-default")
        .await?;

    let session_id = session_id
        .map(|s| s.to_string())
        .unwrap_or_else(coolname::generate_slug);
    let agent_idx = agent_idx.unwrap_or(1);
    let agent_id = format!("{session_id}-{agent_idx}");

    // Build template context
    // Agent Skills spec: global skills live in ~/.agents/skills.
    let skills_dir = dirs::home_dir().expect("home dir").join(".agents/skills");
    let skills = get_skills(&skills_dir);
    let display_tree = build_display_tree(cwd);
    let agents_content = build_agents_content(cwd);

    let prompt_args = serde_json::json!({
        "workspace": cwd,
        "display_tree": display_tree,
        "agents_content": agents_content,
        "session_id": session_id,
        "skills": skills,
    });

    // Render system prompt with handlebars
    let system_prompt = render_template(&template, &prompt_args);

    let request_params = serde_json::json!({"temperature": 0.2});

    // Create agent record
    store
        .create_agent(
            &agent_id,
            &session_id,
            agent_idx,
            cwd,
            &prompt_id,
            &prompt_args,
            &system_prompt,
            tools,
            &request_params,
            model_id,
        )
        .await?;

    let mut session = AgentSession {
        agent_id,
        session_id,
        agent_idx,
        cwd: cwd.to_string(),
        messages: Vec::new(),
        model_identifier: model_id.to_string(),
        tools: tools.clone(),
        request_params,
        prompt_id,
        prompt_args,
    };

    // Start with system message
    let sys_msg = serde_json::json!({"role": "system", "content": system_prompt});
    session.add_message(store, sys_msg, None).await;

    Ok(session)
}

/// Render a Handlebars template with context.
fn render_template(template: &str, args: &serde_json::Value) -> String {
    let mut reg = handlebars::Handlebars::new();
    reg.set_strict_mode(false);
    // Register the template inline
    match reg.render_template(template, args) {
        Ok(rendered) => rendered.trim().to_string(),
        Err(e) => {
            tracing::warn!("template render failed: {e}, using raw");
            // Fallback: basic replacement
            template
                .replace("{{ session_id }}", args.get("session_id").and_then(|v| v.as_str()).unwrap_or(""))
                .replace("{{ workspace }}", args.get("workspace").and_then(|v| v.as_str()).unwrap_or(""))
                .replace("{{ display_tree }}", args.get("display_tree").and_then(|v| v.as_str()).unwrap_or(""))
                .replace("{{ agents_content }}", args.get("agents_content").and_then(|v| v.as_str()).unwrap_or(""))
        }
    }
}

/// Scan skills dir, parse SKILL.md frontmatter, return structured skills.
fn get_skills(skills_dir: &Path) -> Vec<serde_json::Value> {
    let mut skills = Vec::new();
    let Ok(entries) = std::fs::read_dir(skills_dir) else {
        return skills;
    };
    let mut dirs: Vec<_> = entries
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().map(|t| t.is_dir()).unwrap_or(false))
        .collect();
    dirs.sort_by_key(|d| d.file_name());

    for entry in dirs {
        let skill_md = entry.path().join("SKILL.md");
        let Ok(text) = std::fs::read_to_string(&skill_md) else {
            continue;
        };
        if let Some(meta) = parse_frontmatter(&text) {
            let name = meta.get("name").and_then(|v| v.as_str());
            let description = meta.get("description").and_then(|v| v.as_str());
            if let (Some(name), Some(description)) = (name, description) {
                skills.push(serde_json::json!({
                    "name": name.trim(),
                    "description": description.trim(),
                    "path": skill_md.display().to_string(),
                }));
            }
        }
    }
    skills
}

/// Parse YAML frontmatter between --- markers.
fn parse_frontmatter(text: &str) -> Option<serde_json::Value> {
    if !text.starts_with("---") {
        return None;
    }
    let end = text[3..].find("---")?;
    let yaml_str = &text[3..3 + end];
    let val: serde_yaml::Value = serde_yaml::from_str(yaml_str).ok()?;
    serde_json::to_value(val).ok()
}

/// Build directory tree context block.
fn build_display_tree(cwd: &str) -> String {
    let home = dirs::home_dir().unwrap_or_default();
    let home_str = home.display().to_string();
    let notes_dir = home.join(".crow/notes");
    let skills_dir = home.join(".crow/skills");

    let mut trees = Vec::new();
    let t = directory_tree(&notes_dir);
    if !t.is_empty() {
        trees.push(t);
    }
    let t = directory_tree(&skills_dir);
    if !t.is_empty() {
        trees.push(t);
    }
    // Only add cwd tree if cwd != $HOME
    if std::fs::canonicalize(cwd)
        .map(|p| p != std::fs::canonicalize(&home_str).unwrap_or_default())
        .unwrap_or(true)
    {
        let t = directory_tree(Path::new(cwd));
        if !t.is_empty() {
            trees.push(t);
        }
    }
    trees.join("\n\n")
}

/// Build AGENTS.md context block.
fn build_agents_content(cwd: &str) -> String {
    let home = dirs::home_dir().unwrap_or_default();
    let home_str = home.display().to_string();
    let notes_dir = home.join(".crow/notes");
    let mut parts = Vec::new();

    if let Some(content) = read_agents_file(&notes_dir) {
        parts.push(content);
    }
    if std::fs::canonicalize(cwd)
        .map(|p| p != std::fs::canonicalize(&home_str).unwrap_or_default())
        .unwrap_or(true)
    {
        if let Some(content) = read_agents_file(Path::new(cwd)) {
            parts.push(content);
        }
    }
    if parts.is_empty() {
        "No AGENTS.md found".to_string()
    } else {
        parts.join("\n\n")
    }
}

fn read_agents_file(dir: &Path) -> Option<String> {
    for name in ["AGENTS.typ", "AGENTS.md"] {
        let path = dir.join(name);
        if let Ok(content) = std::fs::read_to_string(&path) {
            return Some(content);
        }
    }
    None
}

/// Generate a directory tree string (max depth 3, ignoring common noise).
fn directory_tree(root: &Path) -> String {
    let ignores = [
        "node_modules",
        "__pycache__",
        ".venv",
        "refs",
        "target",
        ".git",
    ];
    let mut lines = Vec::new();
    let root_name = root
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| root.display().to_string());
    lines.push(format!("{root_name}/"));
    walk_tree(root, "", 0, 3, &ignores, &mut lines);
    if lines.len() <= 1 {
        return String::new();
    }
    lines.join("\n")
}

fn walk_tree(
    dir: &Path,
    prefix: &str,
    depth: usize,
    max_depth: usize,
    ignores: &[&str],
    lines: &mut Vec<String>,
) {
    if depth >= max_depth {
        return;
    }
    let mut entries: Vec<_> = match std::fs::read_dir(dir) {
        Ok(e) => e.filter_map(|e| e.ok()).collect(),
        Err(_) => return,
    };
    entries.sort_by_key(|e| {
        let is_dir = e.file_type().map(|t| t.is_dir()).unwrap_or(false);
        (std::cmp::Reverse(is_dir), e.file_name())
    });

    // Filter ignores and hidden files (except .crow)
    entries.retain(|e| {
        let name = e.file_name().to_string_lossy().to_string();
        if ignores.contains(&name.as_str()) {
            return false;
        }
        if name.starts_with('.') && name != ".crow" {
            return false;
        }
        // Skip egg_info patterns
        if name.ends_with(".egg-info") {
            return false;
        }
        true
    });

    for (i, entry) in entries.iter().enumerate() {
        let is_last = i == entries.len() - 1;
        let connector = if is_last { "└── " } else { "├── " };
        let name = entry.file_name().to_string_lossy().to_string();
        let is_dir = entry.file_type().map(|t| t.is_dir()).unwrap_or(false);

        if is_dir {
            lines.push(format!("{prefix}{connector}{name}/"));
            let new_prefix = format!("{prefix}{}", if is_last { "    " } else { "│   " });
            walk_tree(&entry.path(), &new_prefix, depth + 1, max_depth, ignores, lines);
        } else {
            lines.push(format!("{prefix}{connector}{name}"));
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crow_memory_sdk::AgentRecord;

    fn rec(agent_id: &str, idx: i64, created_at: &str) -> AgentRecord {
        AgentRecord {
            agent_id: agent_id.to_string(),
            session_id: "s".to_string(),
            agent_idx: idx,
            cwd: "/tmp".to_string(),
            prompt_id: String::new(),
            prompt_args: serde_json::Value::Null,
            system_prompt: String::new(),
            tool_definitions: serde_json::Value::Null,
            request_params: serde_json::Value::Null,
            model_identifier: "m".to_string(),
            status: "active".to_string(),
            created_at: created_at.to_string(),
        }
    }

    #[test]
    fn resume_picks_highest_idx() {
        let agents = vec![
            rec("s-1", 1, "2026-08-04T12:52:04.000000000+00:00"),
            rec("s-2", 2, "2026-08-04T13:43:24.000000000+00:00"),
        ];
        assert_eq!(pick_resume_agent(&agents).agent_id, "s-2");
    }

    #[test]
    fn system_prompt_template_renders_skills() {
        let template = include_str!("../assets/system_prompt.hbs");
        let args = serde_json::json!({
            "session_id": "cool-name",
            "workspace": "/tmp/ws",
            "display_tree": "tree-here",
            "agents_content": "agents-here",
            "skills": [
                {"name": "plan-todo", "description": "Plan with todos", "path": "/x/plan-todo/SKILL.md"},
            ],
        });
        let rendered = render_template(template, &args);
        assert!(rendered.contains("session id cool-name"));
        assert!(rendered.contains("<SKILLS>"));
        assert!(rendered.contains("**plan-todo** — Plan with todos"));
        assert!(rendered.contains("`/x/plan-todo/SKILL.md`"));
        // No leftover template syntax
        assert!(!rendered.contains("{%"));
        assert!(!rendered.contains("{{"));
        // Content must not be HTML-escaped (triple-mustache used for prose)
        assert!(!rendered.contains("&#x60;"));
    }

    #[test]
    fn system_prompt_template_omits_skills_when_empty() {
        let template = include_str!("../assets/system_prompt.hbs");
        let args = serde_json::json!({
            "session_id": "s",
            "workspace": "/tmp",
            "display_tree": "",
            "agents_content": "",
            "skills": [],
        });
        let rendered = render_template(template, &args);
        // The skills block itself is omitted (the MEMORY section only mentions it in prose)
        assert!(!rendered.contains("You have skills available"));
    }

    #[test]
    fn resume_picks_latest_created_on_idx_tie() {
        // Duplicate generation rows (historical bug: resurrected old runs
        // re-compacted into the same idx) — the newest row is the chain head.
        let agents = vec![
            rec("s-1", 1, "2026-08-04T12:52:04.000000000+00:00"),
            rec("s-2", 2, "2026-08-04T13:43:24.000000000+00:00"),
            rec("s-2", 2, "2026-08-04T14:51:18.000000000+00:00"),
            rec("s-2", 2, "2026-08-04T14:23:51.000000000+00:00"),
        ];
        let picked = pick_resume_agent(&agents);
        assert_eq!(picked.agent_idx, 2);
        assert_eq!(picked.created_at, "2026-08-04T14:51:18.000000000+00:00");
    }

    #[test]
    fn resume_single_record() {
        let agents = vec![rec("s-1", 1, "2026-08-04T12:52:04.000000000+00:00")];
        assert_eq!(pick_resume_agent(&agents).agent_id, "s-1");
    }
}

File: src/crow_cli/agent/session.py
"""
Agent session management - persistence layer for conversation state.

One row = One message. No conv_index gymnastics. No reconstruction headaches.
Just serialize the message dict, deserialize it back.

Agent owns the session. agent_id = "{session_id}-{agent_idx}-{fork_idx}"
is the PK (schema v5; trunk carries fork_idx=1). session_id is derived from
agent_id for ACP upstream routing.
"""

import os
from logging import Logger
from pathlib import Path
from typing import Any

import yaml
from coolname import generate_slug

from crow_cli.agent.memory import DEFAULT_MEMORY_PATH, MemoryClient, MemoryServiceError
from crow_cli.memory import build_agent_id
from crow_cli.agent.prompt import (
    render_template,
    get_directory_tree,
)

from crow_cli.config import (
    AGENTS_DIR,
    Config,
)

async def get_session_by_cwd(cwd, memory_path=DEFAULT_MEMORY_PATH):
    """
    Lookup agents by working directory via the memory db.

    Returns list of session info dicts with session_id, title, updated_at.
    """
    client = MemoryClient(memory_path)
    try:
        result = []
        for agent in await client.list_agents():
            prompt_args = agent.prompt_args or {}
            if prompt_args.get("workspace") != cwd:
                continue

            # Title from the second message (index 1, after the system message).
            # Fetch the first three records so len > 2 tells us a title exists.
            title = "Untitled Chat"
            recs = await client.query_messages(
                agent_id=agent.agent_id, order="asc", limit=3
            )
            if len(recs) > 2:
                try:
                    content = recs[1].data.get("content", "")
                    if isinstance(content, list):
                        title = "".join(
                            b.get("text", "")
                            for b in content
                            if b.get("type") == "text"
                        )[:50]
                    else:
                        title = str(content)[:50]
                    if not title:
                        title = "Untitled Chat"
                except (AttributeError, TypeError):
                    title = "Untitled Chat"

            result.append(
                {
                    "cwd": cwd,
                    "session_id": agent.session_id,
                    "agent_id": agent.agent_id,
                    "title": title,
                    "updated_at": agent.created_at,
                }
            )
        return result
    finally:
        await client.close()


def get_coolname() -> str:
    """Generate a memorable slug"""
    return generate_slug(4)


def snap_turn_cut(messages: list[dict], turn_idx: int | None) -> int | None:
    """Compute the fork cut point for a turn anchor.

    ``turn_idx`` is the 0-based turn (user message) within the agent's
    message list, in the sense "include through the END of this turn". Cut
    points land on user-message boundaries, so an assistant tool_calls group
    is never split from its tool results. Returns the number of messages to
    keep; None keeps everything (fork at HEAD).
    """
    if turn_idx is None:
        return None
    user_idxs = [i for i, m in enumerate(messages) if m.get("role") == "user"]
    if not user_idxs:
        return None
    turn_idx = max(turn_idx, 0)
    if turn_idx + 1 >= len(user_idxs):
        return None  # last turn (or beyond) -> HEAD
    return user_idxs[turn_idx + 1]


async def lookup_or_create_prompt(
    template: str,
    name: str,
    memory_path: str = DEFAULT_MEMORY_PATH,
) -> str:
    """
    Lookup existing prompt by template content, or create new one if not found.
    """
    client = MemoryClient(memory_path)
    try:
        return await client.lookup_or_create_prompt(template, name)
    finally:
        await client.close()


class AgentSession:
    """
    Manages conversation state and persistence via an in-process sqlite db.

    agent_id = "{session_id}-{agent_idx}-{fork_idx}" is the key.
    session_id is derived for ACP upstream routing only.
    """

    def __init__(
        self,
        agent_id: str,
        session_id: str,
        agent_idx: int = 0,
        memory_path: str = DEFAULT_MEMORY_PATH,
        cwd: str = "/tmp",
        fork_idx: int = 1,
        forked_at: str | None = None,
    ):
        self.agent_id = agent_id
        self.session_id = session_id
        self.agent_idx = agent_idx
        self.fork_idx = fork_idx
        self.forked_at = forked_at
        self.memory_path = memory_path
        self.cwd = cwd
        self.messages: list[dict] = []
        self._client = None
        self.model_identifier = None

    @property
    def client(self) -> MemoryClient:
        """Lazy-load the memory client."""
        if self._client is None:
            self._client = MemoryClient(self.memory_path)
        return self._client

    async def add_message(self, msg: dict, usage: dict | None = None):
        """
        Add message to in-memory list AND persist to the service.

        Args:
            msg: Full message dict (role, content, tool_calls, etc.)
            usage: Token usage dict with prompt_tokens, completion_tokens, total_tokens
        """
        self.messages.append(msg)
        await self.client.add_message(self.agent_id, msg, usage)

    async def add_tool_response(
        self,
        tool_results: list[dict],
        logger: Logger,
    ):
        for tool_result in tool_results:
            logger.info(f"TOOL RESULT: {tool_result}")
            await self.add_message(tool_result)

    async def add_assistant_response(
        self,
        thinking: list[str],
        content: list[str],
        tool_call_inputs: list[dict],
        logger: Logger,
        usage: dict | None = None,
    ):
        """
        Handle complex react message building + tool calls + results.

        Args:
            thinking: List of thinking tokens
            content: List of content tokens
            tool_call_inputs: Tool calls from assistant
            logger: Logger instance
            usage: Token usage dict with prompt_tokens, completion_tokens, total_tokens
        """
        # Build react message
        if len(content) > 0 or len(tool_call_inputs) > 0 or len(thinking) > 0:
            thinking_text = "".join(thinking) if thinking else ""
            content_text = "".join(content) if content else ""
            msg = {"role": "assistant", "content": content_text}
            if thinking_text and thinking_text != "":
                msg["reasoning_content"] = thinking_text
            if tool_call_inputs:
                msg["tool_calls"] = tool_call_inputs

            logger.info(f"Adding message: {msg}")
            logger.info(f"Message usage: {usage}")
            # Add to database/list
            await self.add_message(msg, usage)

    async def _save_messages(self, messages: list[dict]):
        """Batch save messages to the service."""
        await self.client.save_messages(self.agent_id, messages)

    @classmethod
    async def create(
        cls,
        prompt_id: str,
        prompt_args: dict[str, Any],
        tool_definitions: list[dict],
        request_params: dict[str, Any],
        model_identifier: str,
        memory_path: str = DEFAULT_MEMORY_PATH,
        cwd: str = "/tmp",
        agent_idx: int = 1,
        fork_idx: int = 1,
        forked_at: str | None = None,
        session_id: str | None = None,
        initial_messages: list[dict[str, Any]] | None = None,
    ) -> "AgentSession":
        """Factory method to create a new agent session."""
        client = MemoryClient(memory_path)

        # Load and render prompt
        try:
            prompt = await client.get_prompt(prompt_id)
        except MemoryServiceError as e:
            if e.status == 404:
                raise ValueError(f"Prompt '{prompt_id}' not found") from e
            raise

        system_prompt = render_template(prompt.template, **prompt_args)
        if session_id is None:
            session_id = get_coolname()
        agent_id = build_agent_id(session_id, agent_idx, fork_idx)

        # Create agent record
        await client.create_agent(
            agent_id=agent_id,
            session_id=session_id,
            agent_idx=agent_idx,
            fork_idx=fork_idx,
            forked_at=forked_at,
            cwd=cwd,
            prompt_id=prompt_id,
            prompt_args=prompt_args,
            system_prompt=system_prompt,
            tool_definitions=tool_definitions,
            request_params=request_params,
            model_identifier=model_identifier,
        )
        await client.close()

        # Build session instance
        session = cls(
            agent_id,
            session_id,
            agent_idx,
            memory_path,
            cwd=cwd,
            fork_idx=fork_idx,
            forked_at=forked_at,
        )
        session.model_identifier = model_identifier
        session.tools = tool_definitions
        session.request_params = request_params
        session.prompt_id = prompt_id
        session.prompt_args = prompt_args

        # Start with system message
        session.messages = [{"role": "system", "content": system_prompt}]
        await session._save_messages(session.messages)

        # Add initial messages if provided
        if initial_messages:
            for msg in initial_messages:
                if msg.get("role") != "system":  # Skip system messages
                    await session.add_message(msg)

        return session

    @classmethod
    async def get_max_agent_idx(
        cls,
        session_id: str,
        memory_path: str = DEFAULT_MEMORY_PATH,
        fork_idx: int | None = 1,
    ) -> int:
        """Return the highest agent_idx for a given session_id.

        fork_idx=1 (default) follows the trunk; None scans all forks.
        """
        client = MemoryClient(memory_path)
        try:
            return await client.get_max_agent_idx(session_id, fork_idx=fork_idx)
        finally:
            await client.close()

    @classmethod
    async def list_sessions(
        cls,
        limit: int = 50,
        offset: int = 0,
        memory_path: str = DEFAULT_MEMORY_PATH,
        include_forks: bool = False,
    ) -> list[dict]:
        """List sessions ordered by most-recent message activity (desc).

        include_forks=False (default) hides the fork dimension: trunk agents
        and their messages only.
        """
        client = MemoryClient(memory_path)
        try:
            return await client.list_sessions(
                limit=limit, offset=offset, include_forks=include_forks
            )
        finally:
            await client.close()

    @classmethod
    async def load(
        cls,
        agent_id: str,
        memory_path: str = DEFAULT_MEMORY_PATH,
    ) -> "AgentSession":
        """Factory method to load existing agent session from the service."""
        client = MemoryClient(memory_path)
        try:
            agent, messages = await client.load(agent_id, hydrate=True)
        except MemoryServiceError as e:
            if e.status == 404:
                raise ValueError(f"Agent '{agent_id}' not found") from e
            raise
        finally:
            await client.close()

        session = cls(
            agent_id=agent.agent_id,
            session_id=agent.session_id,
            agent_idx=agent.agent_idx,
            memory_path=memory_path,
            cwd=agent.cwd,
            fork_idx=agent.fork_idx,
            forked_at=agent.forked_at,
        )
        session.model_identifier = agent.model_identifier
        session.tools = agent.tool_definitions
        session.request_params = agent.request_params
        session.prompt_id = agent.prompt_id
        session.prompt_args = agent.prompt_args
        session.messages = messages

        return session

    @classmethod
    async def fork(
        cls,
        session_id: str,
        memory_path: str = DEFAULT_MEMORY_PATH,
        cwd: str = "/tmp",
        agent_idx: int | None = None,
        turn_idx: int | None = None,
    ) -> "AgentSession":
        """Fork a TRUNK agent of the session (default: HEAD = max agent_idx).

        The fork shares (session_id, agent_idx) with its source and gets the
        next fork_idx. Its context is the trunk's prefix rows up to the turn
        anchor (stored as forked_at, a message-id POSITION) — shared rows,
        never copied. The fork inherits the source's prompt, tools, request
        params and model.
        """
        client = MemoryClient(memory_path)
        try:
            if agent_idx is None:
                agent_idx = await client.get_max_agent_idx(session_id)
            source_id = build_agent_id(session_id, agent_idx, 1)
            source, _ = await client.load(source_id)
            records = await client.query_messages(agent_id=source_id)
            if not records:
                raise ValueError(f"source agent '{source_id}' has no messages to fork")
            cut = snap_turn_cut([r.data for r in records], turn_idx)
            anchor = records[-1].id if cut is None else records[cut - 1].id
            fork_idx = await client.get_max_fork_idx(session_id, agent_idx) + 1
            fork_id = build_agent_id(session_id, agent_idx, fork_idx)
            await client.create_agent(
                agent_id=fork_id,
                session_id=session_id,
                agent_idx=agent_idx,
                fork_idx=fork_idx,
                forked_at=str(anchor),
                cwd=cwd,
                prompt_id=source.prompt_id,
                prompt_args=source.prompt_args,
                system_prompt=source.system_prompt,
                tool_definitions=source.tool_definitions,
                request_params=source.request_params,
                model_identifier=source.model_identifier,
            )
        finally:
            await client.close()

        # Reload through the normal path so the fork's message view (trunk
        # prefix + its own rows) is assembled and hydrated.
        return await cls.load(fork_id, memory_path=memory_path)

    async def close(self):
        """Close the HTTP client."""
        if self._client is not None:
            await self._client.close()
        self._client = None


def _parse_frontmatter(text: str) -> dict | None:
    """Parse a leading YAML frontmatter block (between ``---`` markers).

    Returns the parsed mapping, or None when there is no frontmatter, it is
    unterminated, it is not valid YAML, or it is not a mapping. Uses PyYAML —
    already a project dependency (see configure.py / cli/main.py) — so
    multi-line descriptions, arbitrary indentation, comments, and extra keys
    are all handled for free instead of by a hand-rolled parser.
    """
    if not text.startswith("---"):
        return None
    try:
        end = text.index("---", 3)
        data = yaml.safe_load(text[3:end])
    except (ValueError, yaml.YAMLError):
        return None
    return data if isinstance(data, dict) else None


def get_skills(skills_dir: Path) -> list[dict]:
    """Scan skills dir, parse SKILL.md frontmatter, return structured skills.

    Returns a list of ``{"name", "description", "path"}`` dicts so prompt
    templates can iterate over skills with Jinja (``{% for skill in skills %}``)
    instead of rendering a pre-baked catalog string. ``path`` is the absolute
    path to the skill's SKILL.md so the agent can read it on demand.
    """
    if not skills_dir.exists():
        return []
    skills = []
    for skill_dir in sorted(skills_dir.iterdir()):
        skill_md = skill_dir / "SKILL.md"
        if not skill_md.exists():
            continue
        try:
            text = skill_md.read_text()
        except OSError:
            continue
        meta = _parse_frontmatter(text)
        if not meta:
            continue
        name = meta.get("name")
        description = meta.get("description")
        if name and description:
            skills.append(
                {
                    "name": str(name).strip(),
                    "description": str(description).strip(),
                    "path": str(skill_md),
                }
            )
    return skills


def _read_agents_file(directory: str) -> str | None:
    """Read AGENTS.typ (preferred) or AGENTS.md from a directory.

    Returns None when neither exists, so callers can decide whether to include
    a section rather than emitting a placeholder.
    """
    for name in ("AGENTS.typ", "AGENTS.md"):
        path = os.path.join(directory, name)
        if os.path.exists(path):
            with open(path, "r") as f:
                return f.read()
    return None


def build_display_tree(cwd: str) -> str:
    """Build the directory-tree context block shown to an agent.

    Renders exactly one tree, rooted at cwd — even when cwd is $HOME. Mixing
    the shared ``~/.agents`` workspaces into the tree made models join cwd to
    entries that live elsewhere, producing 404 paths; skills stay discoverable
    through the SKILLS catalog block instead. Returns ``""`` when the tree
    cannot be generated, keeping the ``-> str`` contract.
    """
    return get_directory_tree(cwd)


def build_agents_content(cwd: str) -> str:
    """Build the persistent-memory (AGENTS.md) context block for a session.

    Two memory files are in play:

    * the **global** one at ``~/.agents/AGENTS.md`` — cross-cutting rules
      that carry across every agent, and
    * the **local** one at ``<cwd>/AGENTS.md`` — project-specific knowledge.

    At ``cwd == $HOME`` there is no separate local file (``~/AGENTS.md`` never
    exists — that is all the kicker is for), so only the global memory loads.
    Anywhere else, the global memory loads first and the cwd's own AGENTS.md is
    appended when present.
    """
    home = str(Path.home())
    parts = []
    global_agents = _read_agents_file(str(AGENTS_DIR))
    if global_agents:
        parts.append(global_agents)
    if os.path.realpath(cwd) != os.path.realpath(home):
        local_agents = _read_agents_file(cwd)
        if local_agents:
            parts.append(local_agents)
    return "\n\n".join(parts) if parts else "No AGENTS.md found"


async def make_agent_session(
    config: Config,
    tools: list[dict],
    model_id: str,
    cwd: str,
    session_id: str | None = None,
    agent_idx: int | None = None,
    fork_idx: int = 1,
    forked_at: str | None = None,
):
    if config.system_prompt_path:
        template = config.system_prompt_path.read_text()
    else:
        template = config.system_prompt
    prompt_id = await lookup_or_create_prompt(
        template, name="crow-default", memory_path=config.db_uri
    )
    skills = get_skills(Path(config.skills_dir))

    # Context blocks: the directory tree (cwd only) and the persistent-memory
    # AGENTS.md (global, plus the cwd's own when cwd is not $HOME).
    # See build_display_tree / build_agents_content.
    display_tree = build_display_tree(cwd)
    agents_content = build_agents_content(cwd)
    if session_id is None:
        session_id = get_coolname()
    if agent_idx is None:
        agent_idx = 1
    return await AgentSession.create(
        prompt_id=prompt_id,
        prompt_args={
            "workspace": cwd,
            "display_tree": display_tree,
            "agents_content": agents_content,
            "session_id": session_id,
            "skills": skills,
            "skills_dir": config.skills_dir,
        },
        tool_definitions=tools,
        request_params={"temperature": 0.2},
        model_identifier=model_id,
        memory_path=config.db_uri,
        cwd=cwd,
        agent_idx=agent_idx,
        fork_idx=fork_idx,
        forked_at=forked_at,
        session_id=session_id,
    )
